pkg/v1/partial/uncompressed.go

```diff
@@ -52,29 +56,31 @@ func (ule *uncompressedLayerExtender) Compressed() (io.ReadCloser, error) {
 
 // Digest implements v1.Layer
 func (ule *uncompressedLayerExtender) Digest() (v1.Hash, error) {
-	r, err := ule.Compressed()
-	if err != nil {
-		return v1.Hash{}, err
-	}
-	defer r.Close()
-	h, _, err := v1.SHA256(r)
-	return h, err
+	ule.calcSizeHash()
+	return ule.hash, ule.hashSizeError
 }
 
 // Size implements v1.Layer
 func (ule *uncompressedLayerExtender) Size() (int64, error) {
-	r, err := ule.Compressed()
-	if err != nil {
-		return -1, err
-	}
-	defer r.Close()
-	_, i, err := v1.SHA256(r)
-	return i, err
+	ule.calcSizeHash()
+	return ule.size, ule.hashSizeError
+}
+
+func (ule *uncompressedLayerExtender) calcSizeHash() {
```

Comment:
I think we should still have this return an error, and just return `nil` if the `Once` didn't fire:

```
func calcSizeHash() error {
  var err error
  s.once.Do(func() {
    var r io.ReadCloser
    r, err = ule.Compressed()
    ...
  })
  return err
}
```

This seems cleaner to me than storing the `hashSizeError` on the struct.

Reply:
@ImJasonH I may well be missing something, but won't that be a problem on the second call?  If `ule.Uncompressed` returns an error inside the once then we will have the situation that

upon calling `ule.Digest()` it returns the error that was received from `calcSizeHash`.
But upon the second call, to`url.Size()` we will return nil as the error because the once will not refire, and calcSizeHash() will thus return nil for error.

If you would prefer then I would happily have `calcSizeHash()` store the error in `hashSizeError` and return it so that it could be used, or even return all three values which might be cleaner than what I have now.

I'm open to suggestions, including that I have somehow incorrectly identied the problem

Reply:
Yeah good point. I guess we should keep the same behavior we have today and have `calcSizeHash` return the error every time, regardless of how many times the `Once` is called. I'm convinced.